File: apps/android-tv/app/src/main/java/app/kino/tv/Seasons.kt
package app.kino.tv

import com.stremio.core.types.resource.Video

// The separate unknown-season entry sorts after numbered seasons and Specials.
internal fun Video.seasonNumber() = seriesInfo?.season?.toInt() ?: -1

internal fun availableSeasons(videos: List<Video>) =
    videos.map { it.seasonNumber() }.distinct().sortedBy { if (it < 0) Int.MAX_VALUE else it }

internal fun seasonEpisodes(videos: List<Video>, season: Int) =
    videos
        .filter { it.seasonNumber() == season }
        .sortedWith(
            compareBy<Video> { it.seriesInfo?.episode?.toInt() ?: Int.MAX_VALUE }.thenBy { it.id }
        )

internal fun initialSeason(videos: List<Video>): Int {
    val seasons = availableSeasons(videos)
    val regular = seasons.filter { it > 0 }
    val fallback = if (1 in regular) 1 else regular.firstOrNull() ?: seasons.firstOrNull() ?: -1
    val previous = videos.find { it.currentVideo } ?: return fallback
    if (!previous.watched && (previous.progress ?: 0.0) <= 0) return fallback
    val season = previous.seasonNumber()
    if (season <= 0 || !previous.watched) return season
    // Announced, unreleased episodes still count when identifying the finale.
    val episodes = seasonEpisodes(videos, season)
    val finale = episodes.all { it.seriesInfo != null } && episodes.lastOrNull()?.id == previous.id
    return if (finale) regular.firstOrNull { it > season } ?: season else season
}

File: apps/desktop/src/core/seasons.ts
import type { CoreVideo, LibraryPlaybackProgress } from './types';

export function availableSeasons(videos: CoreVideo[]) {
  return [...new Set(videos.map((video) => video.season))].sort(
    (left, right) => (left ?? Infinity) - (right ?? Infinity),
  );
}

export function seasonEpisodes(videos: CoreVideo[], season: number | null) {
  return videos
    .filter((video) => video.season === season)
    .sort(
      (left, right) =>
        (left.episode ?? Infinity) - (right.episode ?? Infinity) || left.id.localeCompare(right.id),
    );
}

export function initialSeason(videos: CoreVideo[], progress: LibraryPlaybackProgress | null) {
  const seasons = availableSeasons(videos);
  const regular = seasons.filter((season): season is number => season !== null && season > 0);
  const fallback = regular.includes(1) ? 1 : (regular[0] ?? seasons[0] ?? null);
  const previous = videos.find((video) => video.id === progress?.videoId);
  if (!previous || (!previous.watched && !progress?.timeOffset)) return fallback;
  const season = previous.season;
  if (season === null || season === 0 || !previous.watched) return season;
  const episodes = seasonEpisodes(videos, season);
  // Include announced, unreleased episodes. The last released episode alone
  // does not establish the finale of an ongoing season.
  const finale =
    previous.episode !== null &&
    episodes.every((episode) => episode.episode !== null) &&
    episodes.at(-1)?.id === previous.id;
  return finale ? (regular.find((candidate) => candidate > season) ?? season) : season;
}
